custom/auth/ante/burntax.go

```diff
@@ -0,0 +1,61 @@
+package ante
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	treasury "github.com/terra-money/core/x/treasury/types"
+)
+
+// TaxPowerUpgradeHeight is when taxes are allowed to go into effect
+// This will still need a parameter change proposal, but can be activated
+// anytime after this height
+const TaxPowerUpgradeHeight = 9346889
+
+// BurnTaxFeeDecorator will immediately burn the collected Tax
+type BurnTaxFeeDecorator struct {
+	treasuryKeeper TreasuryKeeper
+	bankKeeper     BankKeeper
+}
+
+// NewBurnTaxFeeDecorator returns new tax fee decorator instance
+func NewBurnTaxFeeDecorator(treasuryKeeper TreasuryKeeper, bankKeeper BankKeeper) BurnTaxFeeDecorator {
+	return BurnTaxFeeDecorator{
+		treasuryKeeper: treasuryKeeper,
+		bankKeeper:     bankKeeper,
+	}
+}
+
+// AnteHandle handles msg tax fee checking
+func (btfd BurnTaxFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	// Do not proceed if you are below this block height
+	currHeight := ctx.BlockHeight()
+	if currHeight < TaxPowerUpgradeHeight {
+		return next(ctx, tx, simulate)
+	}
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	msgs := feeTx.GetMsgs()
+
+	// At this point we have already run the DeductFees AnteHandler and taken the fees from the sending account
+	// Now we remove the taxes from the gas reward and immediately burn it
+
+	if !simulate {
+		// Compute taxes again.
+		taxes := FilterMsgAndComputeTax(ctx, btfd.treasuryKeeper, msgs...)
+
+		// Record tax proceeds
+		if !taxes.IsZero() {
+			err = btfd.bankKeeper.SendCoinsFromModuleToModule(ctx, types.FeeCollectorName, treasury.BurnModuleName, taxes)
```

Comment:
Not sure why we would want the BankKeeper to burn on behalf of the Treasury. It would most likely just confuse end-users and there are not really any performance gains associated with it, so could you maybe elaborate on why you feel this would be a good idea?

Reply:
Thank you for clarifying your reasoning :)

Reply:
Thanks for this suggestion.  I originally had the taxes being burned immediately, but then I saw that the taxes are being recorded by the treasury here,
https://github.com/terra-money/classic-core/blob/f2417e1d6a41d3318045210836142e594c84aa79/custom/auth/ante/tax.go#L69
I was thinking it would be more "proper" to let the treasury handle the burn after every block via abci.go.  If the gain in performance warrants this (just saving a module2module send right?) then I can do a direct burn.

Reply:
thank you.  Given that we have performed significant testing using this code, I think we would prefer not to change it.